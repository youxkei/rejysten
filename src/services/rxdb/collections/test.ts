import type { ListItem } from "@/services/rxdb/collections/listItem";

// export type ListItemFixture = [string, ListItemFixture[]] | string;
export type ListItemFixture = [string, ListItemFixture[]?, { text?: string }?];

export function makeListItems(parentId: string, updatedAt: number, fixtures: ListItemFixture[]): ListItem[] {
  const listItems = fixtures.map((fixture) => makeListItem(parentId, updatedAt, fixture));

  for (let i = 0; i < listItems.length; i++) {
    if (i - 1 >= 0) {
      listItems[i][0].prevId = listItems[i - 1][0].id;
    }

    if (i + 1 < listItems.length) {
      listItems[i][0].nextId = listItems[i + 1][0].id;
    }
  }

  return listItems.flat().sort((a, b) => a.id.localeCompare(b.id));
}

function makeListItem(parentId: string, updatedAt: number, [id, children, options]: ListItemFixture): ListItem[] {
  return [{ id, text: options?.text ?? id, prevId: "", nextId: "", parentId, updatedAt }, ...makeListItems(id, updatedAt, children ?? [])];
}
