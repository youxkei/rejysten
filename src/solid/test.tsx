import type { JSXElement } from "solid-js";

import { render } from "@solidjs/testing-library";
import { createEffect, Suspense } from "solid-js";

export async function renderAsync<T extends object>(
  Component: (props: { children: JSXElement }) => JSXElement,
  resolver: (resolve: (value: T) => void) => void
) {
  let resolve: (value: T) => void;
  const promise = new Promise<T>((r) => (resolve = r));

  const result = render(() => (
    <Suspense>
      <Component>
        {(() => {
          createEffect(() => {
            resolver(resolve);
          });

          return null;
        })()}
      </Component>
    </Suspense>
  ));

  const value = await promise;

  return { ...value, ...result };
}
