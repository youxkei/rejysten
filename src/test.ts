const shortener = /^styles_styles_(.+)__\w{8}$/;

export function shortenClassName(root: HTMLElement) {
  for (const element of Array.from(root.querySelectorAll("[class]"))) {
    for (const className of Array.from(element.classList)) {
      const match = className.match(shortener);
      if (match) {
        element.classList.replace(className, match[1]);
      }
    }
  }

  return root;
}

export function getPromiseWithResolve<T = void>() {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  let resolve = (_: T) => {};
  const promise = new Promise<T>((res) => {
    resolve = res;
  });

  return { promise, resolve };
}

export function randomPosInt() {
  return Math.floor(Math.random() * 1024) + 1;
}
