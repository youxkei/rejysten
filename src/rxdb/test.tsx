import type { Collections } from "@/rxdb/collections";
import type { JSX } from "solid-js";

import { indexedDB, IDBKeyRange } from "fake-indexeddb";
import { createRxDatabase, addRxPlugin } from "rxdb";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";
import { RxDBUpdatePlugin } from "rxdb/plugins/update";
import { Suspense } from "solid-js";

import { Provider as CollectionsProvider, collectionCreators } from "@/rxdb/collections";
import { Provider as DatabaseProvider } from "@/rxdb/database";

addRxPlugin(RxDBUpdatePlugin);

export function TestWithRxDB(props: { tid: string; children: JSX.Element }) {
  return (
    <Suspense fallback={props.tid}>
      <DatabaseProvider
        databaseCreator={{
          name: props.tid,
          storage: getRxStorageDexie({
            indexedDB: indexedDB,
            IDBKeyRange: IDBKeyRange,
          }),
          ignoreDuplicate: true,
        }}
      >
        <CollectionsProvider>{props.children}</CollectionsProvider>
      </DatabaseProvider>
    </Suspense>
  );
}

export async function createCollections(tid: string) {
  const database = await createRxDatabase<Collections>({
    name: tid,
    storage: getRxStorageDexie({
      indexedDB: indexedDB,
      IDBKeyRange: IDBKeyRange,
    }),
    ignoreDuplicate: true,
  });

  return database.addCollections(collectionCreators) as Promise<Collections>;
}
