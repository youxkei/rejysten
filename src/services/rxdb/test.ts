import type { Collections } from "@/services/rxdb/collections";

import { indexedDB, IDBKeyRange } from "fake-indexeddb";
import { createRxDatabase } from "rxdb";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";

import { schema } from "@/services/rxdb/collections";

export function createDatabaseCreatorForTest(tid: string) {
  return {
    name: tid,
    storage: getRxStorageDexie({ indexedDB, IDBKeyRange }),
    ignoreDuplicate: true,
  };
}

export async function createCollectionsForTest(tid: string) {
  const database = await createRxDatabase<Collections>({
    name: tid,
    storage: getRxStorageDexie({ indexedDB, IDBKeyRange }),
    ignoreDuplicate: true,
  });

  return database.addCollections(schema) as Promise<Collections>;
}
