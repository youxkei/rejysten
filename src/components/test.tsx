import type { JSXElement } from "solid-js";

import { Suspense } from "solid-js";

import { RxDBServiceProvider } from "@/services/rxdb";
import { createDatabaseCreatorForTest } from "@/services/rxdb/test";

export function TestWithRxDBService(props: { tid: string; children: JSXElement }) {
  return (
    <Suspense fallback={props.tid}>
      <RxDBServiceProvider databaseCreator={createDatabaseCreatorForTest(props.tid)}>
        {props.children}
      </RxDBServiceProvider>
    </Suspense>
  );
}

export { createCollectionsForTest } from "@/services/rxdb/test";
