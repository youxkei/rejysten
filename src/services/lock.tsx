import type { JSXElement } from "solid-js";

import userEvent from "@testing-library/user-event";
import { createContext, createMemo, createSignal, startTransition, untrack, useContext } from "solid-js";

import { renderAsync } from "@/test";

export type LockService = {
  lock$: () => boolean;
  setLock: (lock: boolean) => void;
};

const context = createContext<LockService>();

export function LockServiceProvider(props: { children: JSXElement }) {
  const [lock$, setLock] = createSignal(false);

  return <context.Provider value={{ lock$, setLock }}>{props.children}</context.Provider>;
}

export function useLockService() {
  const service = useContext(context);
  if (!service) throw new Error();

  return service;
}

export async function runWithLock(service: LockService, runner: () => Promise<unknown>) {
  if (untrack(service.lock$)) return;

  service.setLock(true);

  try {
    await runner();
  } finally {
    startTransition(() => service.setLock(false));
  }
}

export function createSignalWithLock<T>(service: LockService, value$: () => T, initialValue: T) {
  const memo$ = createMemo<{ value: T; lock: boolean }>(
    (prev) => {
      const lock = service.lock$();
      if (lock) return { value: prev.value, lock: true };

      return { value: value$(), lock: false };
    },
    { value: initialValue, lock: false },
    { equals: (_, next) => next.lock }
  );

  return () => memo$().value;
}

if (import.meta.vitest) {
  describe("createSignalWithLock and runWithLock", () => {
    test("updated separately without lock", async (test) => {
      const user = userEvent.setup();
      const { container, unmount, findByText } = await renderAsync(
        (props) => (
          <LockServiceProvider>
            {(() => {
              const [x$, setX] = createSignal("x");
              const [y$, setY] = createSignal("y");

              async function onClick() {
                setX("updated x");
                await new Promise((resolve) => queueMicrotask(() => resolve(0)));
                setY("updated y");
              }

              return (
                <>
                  <p>{x$()}</p>
                  <p>{y$()}</p>
                  <button onClick={onClick}>update</button>
                </>
              );
            })()}
            {props.children}
          </LockServiceProvider>
        ),
        (resolve: (value: object) => void) => {
          resolve({});
        }
      );

      test.expect(container).toMatchSnapshot("initial");

      user.click(container.querySelector("button")!);

      await findByText("updated x");
      test.expect(container).toMatchSnapshot("updated x");

      await findByText("updated y");
      test.expect(container).toMatchSnapshot("updated y");

      unmount();
    });

    test("updated separately with createSignalWithLock without runWithLock", async (test) => {
      const user = userEvent.setup();
      const { container, unmount, findByText } = await renderAsync(
        (props) => (
          <LockServiceProvider>
            {(() => {
              const service = useLockService();

              const [x$, setX] = createSignal("x");
              const [y$, setY] = createSignal("y");

              const xWithLock$ = createSignalWithLock(service, x$, "x");
              const yWithLock$ = createSignalWithLock(service, y$, "y");

              async function onClick() {
                setX("updated x");
                await new Promise((resolve) => queueMicrotask(() => resolve(0)));
                setY("updated y");
              }

              return (
                <>
                  <p>{xWithLock$()}</p>
                  <p>{yWithLock$()}</p>
                  <button onClick={onClick}>update</button>
                </>
              );
            })()}
            {props.children}
          </LockServiceProvider>
        ),
        (resolve: (value: object) => void) => {
          resolve({});
        }
      );

      test.expect(container).toMatchSnapshot("initial");

      user.click(container.querySelector("button")!);

      await findByText("updated x");
      test.expect(container).toMatchSnapshot("updated x");

      await findByText("updated y");
      test.expect(container).toMatchSnapshot("updated y");

      unmount();
    });

    test("updated simultaneously with createSignalWithLock and runWithLock", async (test) => {
      const user = userEvent.setup();
      const { container, unmount, findByText } = await renderAsync(
        (props) => (
          <LockServiceProvider>
            {(() => {
              const service = useLockService();

              const [x$, setX] = createSignal("x");
              const [y$, setY] = createSignal("y");

              const xWithLock$ = createSignalWithLock(service, x$, "x");
              const yWithLock$ = createSignalWithLock(service, y$, "y");

              async function onClick() {
                await runWithLock(service, async () => {
                  setX("updated x");
                  await new Promise((resolve) => queueMicrotask(() => resolve(0)));
                  setY("updated y");
                });
              }

              return (
                <>
                  <p>{xWithLock$()}</p>
                  <p>{yWithLock$()}</p>
                  <button onClick={onClick}>update</button>
                </>
              );
            })()}
            {props.children}
          </LockServiceProvider>
        ),
        (resolve: (value: object) => void) => {
          resolve({});
        }
      );

      test.expect(container).toMatchSnapshot("initial");

      user.click(container.querySelector("button")!);

      await findByText("updated x");
      test.expect(container).toMatchSnapshot("updated x");

      unmount();
    });
  });
}
