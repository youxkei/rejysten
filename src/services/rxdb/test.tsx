import type { Collections } from "@/services/rxdb/collections";
import type { JSXElement } from "solid-js";

import { indexedDB, IDBKeyRange } from "fake-indexeddb";
import { createRxDatabase } from "rxdb";
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie";

import { RxDBServiceProvider } from "@/services/rxdb";
import { schema } from "@/services/rxdb/collections";

type Wait = { promise: Promise<Wait> };

export function createDatabaseCreatorForTest(tid: string) {
  return {
    name: tid,
    storage: getRxStorageDexie({ indexedDB, IDBKeyRange }),
    ignoreDuplicate: true,
  };
}

export async function createCollectionsForTest(tid: string) {
  const database = await createRxDatabase<Collections>(createDatabaseCreatorForTest(tid));

  return database.addCollections(schema as any) as Promise<Collections>; // TODO: somehow type check fails so far
}

export function RxDBServiceProviderForTest(props: { tid: string; children: JSXElement }) {
  return <RxDBServiceProvider databaseCreator={createDatabaseCreatorForTest(props.tid)}>{props.children}</RxDBServiceProvider>;
}
