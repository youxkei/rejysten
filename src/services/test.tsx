import type { LockService } from "@/services/lock";
import type { RxDBService } from "@/services/rxdb";
import type { StoreService } from "@/services/store";
import type { JSXElement, Owner } from "solid-js";

import { MultiProvider } from "@solid-primitives/context";
import { Show, createResource, getOwner } from "solid-js";

import { EventServiceProvider } from "@/services/event";
import { EventEmitterServiceProvider } from "@/services/eventEmitter";
import { EventHandlerServiceProvider } from "@/services/eventHandler";
import { createSignalWithLock, runWithLock, LockServiceProvider, useLockService } from "@/services/lock";
import { useRxDBService } from "@/services/rxdb";
import { RxDBServiceProviderForTest } from "@/services/rxdb/test";
import { RxDBSyncFirestoreServiceProvider } from "@/services/rxdbSync/firestore";
import { StoreServiceProvider, useStoreService } from "@/services/store";
import { renderAsync } from "@/solid/test";

type Services = { rxdb: RxDBService; store: StoreService; lock: LockService };

export function renderWithServicesForTest(
  tid: string,
  Component: (props: { children: JSXElement }) => JSXElement,
  setup?: (services: Services) => Promise<unknown>
) {
  return renderAsync(
    (props) => (
      <RxDBServiceProviderForTest tid={tid}>
        <MultiProvider
          values={[
            RxDBSyncFirestoreServiceProvider,
            StoreServiceProvider,
            LockServiceProvider,
            EventServiceProvider,
            EventHandlerServiceProvider,
            EventEmitterServiceProvider,
          ]}
        >
          {(() => {
            if (!setup) {
              return <Component>{props.children}</Component>;
            }

            const rxdb = useRxDBService();
            const store = useStoreService();
            const lock = useLockService();
            const services = { rxdb, store, lock };

            const [done$] = createResource(async () => {
              await runWithLock(lock, () => setup(services));
              return true;
            });

            const doneWithLock$ = createSignalWithLock(lock, () => done$(), false);

            return (
              <Show when={doneWithLock$()}>
                <Component>{props.children}</Component>
              </Show>
            );
          })()}
        </MultiProvider>
      </RxDBServiceProviderForTest>
    ),
    (resolve: (value: { owner: Owner } & Services) => void) => {
      const owner = getOwner()!;

      const rxdb = useRxDBService();
      const store = useStoreService();
      const lock = useLockService();

      resolve({ owner, rxdb, store, lock });
    }
  );
}
