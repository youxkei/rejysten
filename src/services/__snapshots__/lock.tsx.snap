// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSignalWithLock and runWithLock > updated separately with createSignalWithLock without runWithLock > initial 1`] = `
<div>
  <p>
    x
  </p>
  <p>
    y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated separately with createSignalWithLock without runWithLock > updated x 1`] = `
<div>
  <p>
    updated x
  </p>
  <p>
    y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated separately with createSignalWithLock without runWithLock > updated y 1`] = `
<div>
  <p>
    updated x
  </p>
  <p>
    updated y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated separately without lock > initial 1`] = `
<div>
  <p>
    x
  </p>
  <p>
    y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated separately without lock > updated x 1`] = `
<div>
  <p>
    updated x
  </p>
  <p>
    y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated separately without lock > updated y 1`] = `
<div>
  <p>
    updated x
  </p>
  <p>
    updated y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated simultaneously with createSignalWithLock and runWithLock > initial 1`] = `
<div>
  <p>
    x
  </p>
  <p>
    y
  </p>
  <button>
    update
  </button>
</div>
`;

exports[`createSignalWithLock and runWithLock > updated simultaneously with createSignalWithLock and runWithLock > updated x 1`] = `
<div>
  <p>
    updated x
  </p>
  <p>
    updated y
  </p>
  <button>
    update
  </button>
</div>
`;
